### DISCUSSION TITLE
Assign permissions to Pivot Table

### DISCUSSION THREAD

**Discussion by sys-auditing** [spatie/laravel-permission#1782] [Q&A]
Hi, how to Use Spatie Laravel-permission Package for Pivot Table ?

**Comment by erikn69** [comment]
You just have to add a pivot on the tables "roles" and "model_has_permissions", and then overwrite the relationship in the User model "public function roles(): MorphToMany" and "permissions(): MorphToMany", i use session for save the new pivot on login, and later i use the session function for add the pivot on user relationship with whereraw, is that what whas loking for?

**Comment by kaaaaaaaaaaai** [comment]
@erikn69 
I'm stuck with the same problem.
Could you please tell me more about it?

**Comment by erikn69** [comment]
@kaaaaaaaaaaai  add a pivot field, and on login authenticated i put that id on session
![image](https://user-images.githubusercontent.com/4933954/124485150-e5b7c100-dd71-11eb-9258-dc396b84d88d.png)

Now overwrite relations on user model and add a where for session id
**ROLES:**
https://github.com/spatie/laravel-permission/blob/a6e4122b65094baba7f98df153af0768ef910c85/src/Traits/HasRoles.php#L37-L49
```php
public function roles(): BelongsToMany
    {
        return $this->morphToMany(
            config('permission.models.role'),
            'model',
            config('permission.table_names.model_has_roles'),
            config('permission.column_names.model_morph_key'),
            'role_id'
        )->where(config('permission.table_names.model_has_roles').'.empr_id',session('empr_id'));
    }
```
**Permissions**
https://github.com/spatie/laravel-permission/blob/a6e4122b65094baba7f98df153af0768ef910c85/src/Traits/HasPermissions.php#L40-L52
```php
public function permissions(): BelongsToMany 
     { 
         return $this->morphToMany( 
             config('permission.models.permission'), 
             'model', 
             config('permission.table_names.model_has_permissions'), 
             config('permission.column_names.model_morph_key'), 
             'permission_id' 
         )->where(config('permission.table_names.model_has_permissions').'.empr_id',session('empr_id')); 
     } 
```
Now works with a pivot id based on session, with that example also you can made custom roles for each pivot id

**Comment by erikn69** [comment]
And maybe duplicate of [#1634](https://github.com/spatie/laravel-permission/issues/1634)

**Comment by kaaaaaaaaaaai** [comment]
@erikn69 Thanks for the details, I tried writing your code to see if it worked, but it didn't work with my understanding. Can you help me?Would it work in my situation?

My situation is
There are communities and users, and communities are divided by subdomains.

<img width="455" alt="スクリーンショット 2021-07-10 11 25 00" src="https://user-images.githubusercontent.com/2194021/125149079-b5a55f00-e171-11eb-9fe1-78d3013c16d4.png">

ex.)
User A has the Writer role when the subdomain is "sample
When the subdomain is "test", the owner role

User B, when the subdomain is "sample", has the Writer role
I would like to grant
(The permission is omitted.)

I'm hoping for a result like this.
```
//user A loggined and browing "sample.domain.com"
Auth::user()->getRoleNames();
//["writer"]
```

**Comment by erikn69** [comment]
@kaaaaaaaaaaai  potato patato, the same with diferent ids
![image](https://user-images.githubusercontent.com/4933954/125310469-39895380-e2f8-11eb-8d1e-b959b62582ce.png)
on login authenticated, put that id on session with `session(['community_id'=>$communityModel->id]);`, 
(i am assuming that you already have a multi community login, and maybe custom session guard for support _remember_me_ behaivor with `community_id`, `\Illuminate\Auth\Recaller`)
Also you can add `community_id` pivot on `roles` if you wants custom private roles for a community, and when that id is null, the role is global

**Then**, all the code on user
```php
public function roles(): BelongsToMany
{
        return $this->morphToMany(
            config('permission.models.role'),
            'model',
            config('permission.table_names.model_has_roles'),
            config('permission.column_names.model_morph_key'),
            'role_id'
        )->where(config('permission.table_names.model_has_roles').'.community_id',session('community_id'));
}
public function permissions(): BelongsToMany 
{ 
         return $this->morphToMany( 
             config('permission.models.permission'), 
             'model', 
             config('permission.table_names.model_has_permissions'), 
             config('permission.column_names.model_morph_key'), 
             'permission_id' 
         )->where(config('permission.table_names.model_has_permissions').'.community_id',session('community_id')); 
} 
```
Now, User A can has the Writer role when the subdomain is "sample", and when the subdomain is "test", the owner role, and repeat the same role or has diferents on other subdomain because `community_id` is the new pivot and composite primary key

Also, in addition maybe you have to overwrite `public function assignRole(...$roles){}` and `public function givePermissionTo(...$permissions){}` changing `->sync($roles or $permissions, false)` for  `->syncWithPivotValues($roles or $permissions, ['community_id'=>session('community_id')], false)`  if you wants to change permissions from your application instead of manually

It can be improved, but it works, nobody wants to do a PR for this, nor do they provide information, maybe @freekmurze can better understand the idea and add it to the package

Also using `model_type` can be another way, adding more models for user's table, and on config `auth.providers` more classes and guards, so you can have a new model guard for every community, but then you have N+1 issue

**Comment by fredsal** [comment]
@kaaaaaaaaaaai i have the same problem, did erikn69's solution works for you?
@drbyte is there a better way to achieve this?

**Comment by drbyte** [comment]
Yes, overriding the trait's relations in your app is a suitable way to accomplish this.
